pg_replicate/src/pipeline/sinks/delta.rs

```diff
@@ -0,0 +1,143 @@
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+};
+
+use async_trait::async_trait;
+use tokio_postgres::types::PgLsn;
+use tracing::info;
+
+use super::{BatchSink, SinkError};
+use crate::{
+    clients::delta::DeltaClient,
+    conversions::{cdc_event::CdcEvent, table_row::TableRow},
+    pipeline::PipelineResumptionState,
+    table::{TableId, TableSchema},
+};
+use deltalake::arrow::error::ArrowError;
+use deltalake::{arrow::datatypes::Schema, DeltaTableError};
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum DeltaSinkError {
+    #[error("Arrow error: {0}")]
+    DeltaArrow(#[from] ArrowError),
+    #[error("delta error: {0}")]
+    Delta(#[from] DeltaTableError),
+    #[error("missing table schemas")]
+    MissingTableSchemas,
+    #[error("missing delta schemas")]
+    MissingDeltaSchemas(String),
+    #[error("missing table id: {0}")]
+    MissingTableId(TableId),
+    #[error("incorrect commit lsn: {0}(expected: {0})")]
+    IncorrectCommitLsn(PgLsn, PgLsn),
+    #[error("commit message without begin message")]
+    CommitWithoutBegin,
+}
+
+pub struct DeltaSink {
+    client: DeltaClient,
+}
+
+impl DeltaSink {
+    pub fn new(path: String) -> Self {
+        DeltaSink {
+            client: DeltaClient {
+                path,
+                table_schemas: None,
+                delta_schemas: None,
+            },
+        }
+    }
+}
+
+impl SinkError for DeltaSinkError {}
+
+#[async_trait]
+impl BatchSink for DeltaSink {
+    type Error = DeltaSinkError;
+    async fn get_resumption_state(&mut self) -> Result<PipelineResumptionState, Self::Error> {
+        Ok(PipelineResumptionState {
+            copied_tables: HashSet::new(),
+            last_lsn: PgLsn::from(0),
+        })
+    }
+
+    async fn write_table_schemas(
+        &mut self,
+        table_schemas: HashMap<TableId, TableSchema>,
+    ) -> Result<(), Self::Error> {
+        let mut delta_schema: HashMap<String, Arc<Schema>> = HashMap::new();
+
+        for (_, table_schema) in &table_schemas {
```

Comment:
This for loop iterates over the hashmap's values so can probably be rewritten as `for table_schema in table_schemas.values()`